Detect conflicts between exported types and FQNs

hubertp opened the pull request:
### Pull Request Description

Exporting types named the same as the module where they are defined in `Main` modules of library components may lead to accidental name conflicts. This became apparent when trying to access `Problem_Behavior` module via a fully qualified name and the compiler rejected it. This is due to the fact that `Main` module exported `Error` type defined in `Standard.Base.Error` module, thus making it impossible to access any other submodules of `Standard.Base.Error` via a fully qualified name.

This change adds a warning to FullyQualifiedNames pass that detects any such future problems.
While only `Error` module was affected, it was widely used in the stdlib, hence the number of changes.

Closes #5902.

### Important Notes

I left out the potential conflict in micro-distribution, thus ensuring we actually detect and report the warning.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] ~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] ~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~

hubertp committed:
Detect conflicts between exported types and FQNs

Exporting types named the same as the module where they are defined in
`Main` modules of library components may lead to accidental name
conflicts. This became apparent when trying to access `Problem_Behavior`
module via a fully qualified name and the compiler rejected it.
This is due to the fact that `Main` module exported `Error` type defined
in `Standard.Base.Error` module, thus making it impossible to access any
other submodules of `Standard.Base.Error` via a fully qualified name.

This change adds a warning to FullyQualifiedNames pass that detects any
such future problems.
While only `Error` module was affected, it was widely used in the
stdlib, hence the number of changes.

I left out the potential conflict in micro-distribution, thus ensuring
we actually detect and report the warning.

Closes #5902.

hubertp committed:
update changelog

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:
`````diff
@@ -680,7 +680,7 @@ class Compiler(
     * ````
     *   type C
-    *       type C a
+    *       C a
`````
If we can catch this now, then great!

hubertp replied:
This documentation was using the outdated syntax

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso:
Can we move the `throw` method up from the type to the module so we keep `Error.throw` please.

hubertp replied:
We can, sure. Although I understood that this would be deprecated eventually.

jdunkerley replied:
No we use `Error.throw` everywhere - we are minimising panics but we use errors all the time.

radeusgd replied:
And I think while we are minimising panics, we are not planning to deprecate them either, right? We don't want to expose them to users, but they are a necessary part of internal error handling.

hubertp replied:
Actually, it's not as straightforward as it seems. This is because, again, of the name conflict with the exported `project.Error` in `Main.enso`

jdunkerley replied:
how about the following then ...
Lets rename `Standard.Base.Error.*` to `Standard.Base.Errors.*` (Common etc...) and keep `Standard.Base.Error.Error` (i.e. the Error.enso file) as is,

I feel that is the cleaner?

hubertp replied:
> And I think while we are minimising panics, we are not planning to deprecate them either, right? We don't want to expose them to users, but they are a necessary part of internal error handling.

I was referring to having (static) module methods.

jdunkerley replied:
ah ok - I thought you were meaning throwing errors :)

hubertp replied:
done

hubertp committed:
nit

hubertp committed:
s/Error/Errors

hubertp committed:
s/Error/Errors part 2

hubertp committed:
Added a standalone test

hubertp committed:
micro-distribution renamings

hubertp committed:
fmt

hubertp committed:
Synthetic modules need their bindings restored

We don't serialize/deserialize IR for synthetic modules so with the new
approach, where bindings are being restored from caches, they would be
empty for synthetic modules.
This bug got revealed now because we didn't have a test that checks
FQN access involving such a synthetic module. This would have bitten us
sooner than later.

TODO for future - consider serializing IRs for synthetic modules.

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/5902-detect-name-conflicts-with-fqns
